### PR TITLE
ZoneHVAC:UnitVentilator shows different autosized min/max OA flow rates for FixedAmount OA control option

### LINKED ISSUE #9284: ZoneHVAC:UnitVentilator shows confusing autosized min/max outdoor air values in sizing results
Issue overview
--------------
The unit ventilator has 3 OA control options. The FixedAmount choice will maintain a constant flow rate, the others will change between min and max based on RA to OA temperature difference.

   A3 , \field Outdoor Air Control Type
        \required-field
        \type choice
        \key FixedAmount
        \key VariablePercent
        \key FixedTemperature

When choosing FixedAmount (Zone 1 in defect file) the eio and html shows the max outdoor air different from the min outdoor air in several reports.

The eio file shows:

    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.24063
    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Minimum Outdoor Air Flow Rate [m3/s], 0.10384

The Coil Sizing Details shows:

image

While results show that the unit ventilator outdoor air flow rate is about 50% of max air flow rate.

image

The values reported for FixedAmount should be identical if that is how the simulation controls outdoor air.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added 
([UnitVent5ZoneAuto_wCoilSizing.idf.txt](https://github.com/NREL/EnergyPlus/files/8100625/UnitVent5ZoneAuto_wCoilSizing.idf.txt))
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes #9284
 - Correct max OA flow rate autosizing for FixedAmount OA control option in ZoneHVAC:UnitVentilator.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/UnitVentilator.unit.cc [hunk 1]
```diff
@@ -272,4 +273,80 @@ TEST_F(EnergyPlusFixture, UnitVentilatorCalcMdotCCoilCycFanTest)
...
+    // retest without HVAC sizing object
+    state->dataUnitVentilators->UnitVent(UnitVentNum).HVACSizingIndex = 0;
+
+    // test fixed OA control
+    state->dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow = DataSizing::AutoSize;
+    state->dataUnitVentilators->UnitVent(UnitVentNum).OAControlType = fixedOAControl;
+    UnitVentilator::SizeUnitVentilator(*state, UnitVentNum);
+
+    EXPECT_NEAR(
+        state->dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow, state->dataUnitVentilators->UnitVent(UnitVentNum).MaxAirVolFlow, 0.001);
```
Comment (rraustad) [inline tst/EnergyPlus/unit/UnitVentilator.unit.cc]: This result should match minOA.

File: src/EnergyPlus/UnitVentilator.cc [hunk 2]
```diff
@@ -2224,7 +2224,24 @@ namespace UnitVentilator {
...
+                    if (state.dataUnitVentilators->UnitVent(UnitVentNum).HVACSizingIndex > 0) {
+                        // use value set above
+                        OutAirVolFlowDes = ZoneEqSizing(state.dataSize->CurZoneEqNum).OAVolFlow;
+                    } else if (state.dataSize->FinalZoneSizing(state.dataSize->CurZoneEqNum).ZoneDesignSpecOAIndex > 0) {
+                        // the user has specified inputs for OA flow
+                        OutAirVolFlowDes = min(state.dataSize->FinalZoneSizing(state.dataSize->CurZoneEqNum).MinOA,
+                                               state.dataUnitVentilators->UnitVent(UnitVentNum).MaxAirVolFlow);
+                    } else {
+                        OutAirVolFlowDes = state.dataUnitVentilators->UnitVent(UnitVentNum).MaxAirVolFlow;
+                    }
```
Comment (rraustad) [inline src/EnergyPlus/UnitVentilator.cc]: Seems this else if and else should be combined into an else to yield minOA.

**Comment by rraustad** [comment]
Defect file results:

    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.10384
    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Minimum Outdoor Air Flow Rate [m3/s], 0.10384

image

image

image

File: src/EnergyPlus/UnitVentilator.cc [hunk 3]
```diff
@@ -2224,7 +2224,16 @@ namespace UnitVentilator {
                 }
             } else {
                 CheckZoneSizing(state, state.dataUnitVentilators->cMO_UnitVentilator, state.dataUnitVentilators->UnitVent(UnitVentNum).Name);
-                OutAirVolFlowDes = state.dataUnitVentilators->UnitVent(UnitVentNum).MaxAirVolFlow;
```
Comment (rraustad) [inline src/EnergyPlus/UnitVentilator.cc:2227]: If OA Control = FixedAmount match the min OA for reporting purposes.

**Comment by rraustad** [comment]
This diff in DOAToUnitVentilator implies that both max and min OA flow should be set to 0 when an AT mixer is used. The eio coil sizing did not change. The eso results did not change. This tells me, so far, this type of change is OK.

      Component Sizing Information, ZoneHVAC:UnitVentilator, TU1, Design Size Maximum Supply Air Flow Rate [m3/s], 0.24020
    - Component Sizing Information, ZoneHVAC:UnitVentilator, TU1, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.24020
    + Component Sizing Information, ZoneHVAC:UnitVentilator, TU1, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.00000
      Component Sizing Information, ZoneHVAC:UnitVentilator, TU1, Design Size Minimum Outdoor Air Flow Rate [m3/s], 0.10384

**Comment by rraustad** [comment]
There is one more change to implement here (min OA = 0 when using AT mixers). I will wait for comments before making that change.

File: src/EnergyPlus/UnitVentilator.cc [hunk 4]
```diff
@@ -2224,7 +2224,13 @@ namespace UnitVentilator {
...
+                if (state.dataUnitVentilators->UnitVent(UnitVentNum).OAControlType == FixedOAControl) {
```
Comment (amirroth) [inline src/EnergyPlus/UnitVentilator.cc:2227]: Fun new pattern. Use `auto &unitVent = state.dataUnitVentilators->UnitVent(UnitVentNum);` and then `unitVent` to shorten lines and make code more readable.

Comment (rraustad) [reply src/EnergyPlus/UnitVentilator.cc:2227]: That would mask this change by changing many lines and then this one change is hard to see (unless of course I commented that this is the only change). Once the correction is agreed upon, yes, I can make that change.

Comment (amirroth) [reply src/EnergyPlus/UnitVentilator.cc:2227]: 100% understand. Just a suggestion for the future.

**Comment by rraustad** [comment]
Plan B - setting the OA flow = 0 when AT mixer is present turns off the DOAS system. So the only change is to correctly account for the FixedAmount OA control type. I don't think any example files use this control type so am expecting 0 diffs. The DOAToUnitVentilator files now looks the same before and after this fix.

image

**Comment by rraustad** [comment]
The defect file results also looks the same before and after except now reports max OA flow correctly. Documentation shows that the FixedAmount OA control does use the minimum OA flow rate and schedule, however, this bait and switch happens during getInput and if min OA field is autosized then this same logic has to be accounted for in sizing.

    if (state.dataUnitVentilators->UnitVent(UnitVentNum).OAControlType == FixedOAControl) {
        state.dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow = state.dataUnitVentilators->UnitVent(UnitVentNum).MinOutAirVolFlow;
        state.dataUnitVentilators->UnitVent(UnitVentNum).MaxOASchedName = state.dataUnitVentilators->UnitVent(UnitVentNum).MinOASchedName;
        state.dataUnitVentilators->UnitVent(UnitVentNum).MaxOASchedPtr =
            GetScheduleIndex(state, state.dataUnitVentilators->UnitVent(UnitVentNum).MinOASchedName);
    }

    ZoneHVAC:UnitVentilator,
      Zone1UnitVent,           !- Name
      UnitVentAvailability,    !- Availability Schedule Name
      autosize,                !- Maximum Supply Air Flow Rate {m3/s}
      FixedAmount,             !- Outdoor Air Control Type
      autosize,                !- Minimum Outdoor Air Flow Rate {m3/s}
      UnitVentMinOA,           !- Minimum Outdoor Air Schedule Name
      autosize,                !- Maximum Outdoor Air Flow Rate {m3/s}

Develop:

    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.24063
    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Minimum Outdoor Air Flow Rate [m3/s], 0.10384

This branch:

    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Maximum Outdoor Air Flow Rate [m3/s], 0.10384
    Component Sizing Information, ZoneHVAC:UnitVentilator, ZONE1UNITVENT, Design Size Minimum Outdoor Air Flow Rate [m3/s], 0.10384

image

**Comment by rraustad** [comment]
@mjwitte can you take a quick look at this? I have another commit for lots of refactoring that will muddy the waters.

File: src/EnergyPlus/UnitVentilator.cc (hunk 3 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitVentilator.cc:2227]: So with this change, the unit ventilator will now respect the FinalZoneSizing MinOA value, and only if the OA control type is fixed OA control.  Interesting.

File: testfiles/UnitVent5ZoneFixedOANoCoilOpt.idf [hunk 5]
```diff
@@ -2,7 +2,7 @@
 !-Option OriginalOrderTop UseSpecialFormat
 !-NOTE: All comments with '!-' are ignored by the IDFEditor and are generated automatically.
 !-      Use '!' comments if they need to be retained when using the IDFEditor.
-! 5ZoneAirCooled.idf
```
Comment (Myoldmopar) [inline testfiles/UnitVent5ZoneFixedOANoCoilOpt.idf:5]: Obviously no problem with these fixups, but you can also just keep a tally of papercut issues like this and fix them in one non-functional-change PR so they don't interfere.  In this case, obviously it doesn't interfere, just noting it.

File: tst/EnergyPlus/unit/UnitVentilator.unit.cc [hunk 6]
```diff
@@ -272,4 +273,81 @@ TEST_F(EnergyPlusFixture, UnitVentilatorCalcMdotCCoilCycFanTest)
...
+
+    // test fixed temperature OA control
+    state->dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow = DataSizing::AutoSize;
+    state->dataUnitVentilators->UnitVent(UnitVentNum).OAControlType = fixedTemperature;
+    UnitVentilator::SizeUnitVentilator(*state, UnitVentNum);
+
+    EXPECT_NEAR(
+        state->dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow, state->dataUnitVentilators->UnitVent(UnitVentNum).MaxAirVolFlow, 0.001);
+}
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/UnitVentilator.unit.cc:351]: This is a great unit test.  Testing several different permutations.  Good stuff!

**Review by Myoldmopar — APPROVED** [review]
Yeah no problem here.  There are others that are more well-versed on the implications of the OA/sizing interactions, so I will be relying on the plots and results shown on the PR.  The change itself makes sense, and the unit test is great.

**Comment by Myoldmopar** [comment]
CI is all clean, and the sizing results you show above make sense for fixed OA control.  I'm OK with this going in, but open to others to review it before that.

File: src/EnergyPlus/UnitVentilator.cc (hunk 3 above)
Comment (rraustad) [reply src/EnergyPlus/UnitVentilator.cc:2227]: Correct. In getInput the model uses the minimum flow for FixedAmount control. But when autosizing these air vol flows, the sizing routine forgot about that requirement.

            if (state.dataUnitVentilators->UnitVent(UnitVentNum).OAControlType == state.dataUnitVentilators->FixedOAControl) {
                state.dataUnitVentilators->UnitVent(UnitVentNum).OutAirVolFlow = state.dataUnitVentilators->UnitVent(UnitVentNum).MinOutAirVolFlow;
                state.dataUnitVentilators->UnitVent(UnitVentNum).MaxOASchedName = state.dataUnitVentilators->UnitVent(UnitVentNum).MinOASchedName;
                state.dataUnitVentilators->UnitVent(UnitVentNum).MaxOASchedPtr =
                    GetScheduleIndex(state, state.dataUnitVentilators->UnitVent(UnitVentNum).MinOASchedName);
            }

**Comment by Myoldmopar** [comment]
Last chance for additional commentary.  I'm building with latest develop just to verify it's good now.

**Comment by Myoldmopar** [comment]
It was all happy with develop pulled in, merging this and moving on.  Thanks @rraustad

**Pull request merged**